### ISSUE TITLE
False positive InvalidCast on key-of<array<array-key, mixed>>

### ISSUE THREAD

**Issue by VincentLanglet** [vimeo/psalm#10253]
Reproducer https://psalm.dev/r/65f20eade6

Am I wrong or `key-of<array<array-key, mixed>>` is `array-key` ?
And so I should be able to cast it to string ?

**Comment by robchett** [comment]
```php 
/**
 * @param array<string, array<string, mixed>> $i
 */
function takesAnInt(array $i, string $key): mixed {
    $ikey = $i[$key]; // array<string, mixed>|null
    $fields = $i[$key]['fields']; // mixed
    $keys = array_keys($fields); // ????
    foreach ($keys as $field) {
        $a = (string) $field;
    }
    
    return $a ?? '';
}
```

perhaps your @param definition should be `@param array<string, array{fields: array<string, mixed>}> $i`?

https://psalm.dev/r/35aa3516b6

**Comment by VincentLanglet** [comment]
> perhaps your @param definition should be `@param array<string, array{fields: array<string, mixed>}> $i`?

The real use case if more complex ; I tried to simplify.

Sure I can solve my issue with another phpdoc/code, like https://psalm.dev/r/e2758d81a6 which doesn't report the InvalidCast anymore.

Still, I think something is wrong in psalm codebase because 
```
key-of<array<array-key, mixed>>
```
should be castable to string.

**Comment by robchett** [comment]
Ok, I've produced a simpler repo for this - https://psalm.dev/r/e3d5d1f809.
Essentially `array_keys(mixed)` has a weird return type because it is expecting an array and doesn't really know what to do with `list<key-of<mixed>>`, leading to unclear error messaging.

```php
  /**
   * @psalm-template TArray as array
   *
   * @param TArray $array
   * @param mixed $search_value
   * @param bool  $strict
   *
   * @return (TArray is non-empty-array ? non-empty-list<key-of<TArray>> : list<key-of<TArray>>)
   * @psalm-pure
 */
```

Its a `typeError` to pass a non-array to `array_keys` so a return type provider that converts `mixed -> list<array-key>` should be safe
